scripts/cn-setup.sh

```diff
@@ -7,8 +7,29 @@ ulimit -n 65535
 
 TF=/opt/starrocks/cn/conf/cn.conf
 
+# Make ports configurable
 sed -i "s#^be_port =.*#be_port = \${THRIFT_RPC_PORT}#" $TF
 sed -i "s#^be_http_port =.*#be_http_port = \${HTTP_PORT}#" $TF
 sed -i "s#^heartbeat_service_port =.*#heartbeat_service_port = \${HEARTBEAT_PORT}#" $TF
 sed -i "s#^brpc_port =.*#brpc_port = \${BRPC_PORT}#" $TF
 sed -i "s#^starlet_port =.*#starlet_port = \${STARLET_PORT}#" $TF
+
+# Set up data config
+echo 'block_cache_disk_path=${STARROCKS_DATA}/block_cache/' >> $TF
+echo 'block_cache_meta_path=${STARROCKS_DATA}/block_cache/' >> $TF
+echo 'datacache_disk_path=${STARROCKS_DATA}/datacache/' >> $TF
+echo 'datacache_meta_path=${STARROCKS_DATA}/datacache/' >> $TF
+echo 'user_function_dir=${STARROCKS_DATA}/lib/udf' >> $TF
+echo 'local_library_dir=${STARROCKS_DATA}/lib/udf-runtime' >> $TF
+echo 'pprof_profile_dir=${STARROCKS_DATA}/log' >> $TF
+echo 'query_scratch_dirs=${STARROCKS_DATA}' >> $TF
+echo 'small_file_dir=${STARROCKS_DATA}/lib/small_file/' >> $TF
+echo 'spill_local_storage_dir=${STARROCKS_DATA}/spill' >> $TF
+echo 'starlet_cache_dir=${STARROCKS_DATA}/storage/starlet_cache' >> $TF
+echo 'storage_root_path=${STARROCKS_DATA}/storage' >> $TF
+echo 'sys_log_dir=${STARROCKS_DATA}/log' >> $TF
+echo 'sys_minidump_dir=${STARROCKS_DATA}' >> $TF
+
+# I frankly have no idea what these are or whether they need to be persisted
+echo 'pull_load_task_dir=${STARROCKS_DATA}/var/pull_load' >> $TF
+echo 'query_debug_trace_dir=${STARROCKS_DATA}/query_debug_trace' >> $TF
```

Comment:
If we don't explicitly set those sub-dir path, will the service automatically create them using a default value? What happen if we forget to config one of them? They are quite many sub-dir path we need to config manually.

Reply:
Maybe also add a reference to explain why/what sub-dir we need to config.

Reply:
> If we don't explicitly set those sub-dir path, will the service automatically create them using a default value?

Yes, replace `${STARROCKS_DATA}` with `/opt/starrocks` and that's the path

> What happen if we forget to config one of them? They are quite many sub-dir path we need to config manually.

The data is created inside the container instead of the volume, which means when we make a new container (say for an upgrade) the data is gone.

And yeah, this solution sucks, but for now it seems to be the only way. At least the StarRocks Slack hasn't had any better suggestions so far.

Reply:
👍 hopefully we didn't miss any sub-dirs.

Reply:
> Maybe also add a reference to explain why/what sub-dir we need to config.

I'm not sure what you mean by this, can you elaborate?

Reply:
All the subdirs for the FE and CN are listed in the variables section of the UI. I just grabbed anything that started with `/opt/starrocks` and didn't look like a place to put library code like jars, etc.

Reply:
Where did you get all of those settings? Is there a StarRocks document that people can look at?

Reply:
From the CN and FE UIs

Reply:
👍 That's fine.

Reply:
https://github.com/StarRocks/starrocks/issues/47049

Reply:
💯